From f804b2b098d6611db2493a4833e6935e08175ec0 Mon Sep 17 00:00:00 2001
From: Alastair D'Silva <alastair@d-silva.org>
Date: Sat, 15 Aug 2026 20:21:11 +1000
Subject: [PATCH 04/23] clk: sunxi: Add Allwinner A733 CCU and R-CCU clock
 drivers

Add clock gate and reset control definitions for the main CCU and R-CCU
on Allwinner A733 / T736 (sun60iw2).

Follows the driver model established in clk_a523.c by Andre Przywara,
with gate and reset register offsets ported from Allwinner's BSP
ccu-sun60iw2.c driver.

Signed-off-by: Alastair D'Silva <alastair@d-silva.org>
Assisted-by: Antigravity <antigravity@google.com>
---
 drivers/clk/sunxi/Kconfig      | 14 +++++
 drivers/clk/sunxi/Makefile     |  2 +
 drivers/clk/sunxi/clk_a733.c   | 94 ++++++++++++++++++++++++++++++++++
 drivers/clk/sunxi/clk_a733_r.c | 38 ++++++++++++++
 drivers/clk/sunxi/clk_sunxi.c  | 10 ++++
 5 files changed, 158 insertions(+)
 create mode 100644 drivers/clk/sunxi/clk_a733.c
 create mode 100644 drivers/clk/sunxi/clk_a733_r.c

diff --git a/drivers/clk/sunxi/Kconfig b/drivers/clk/sunxi/Kconfig
index 1c1cc82719c..9858f4546ea 100644
--- a/drivers/clk/sunxi/Kconfig
+++ b/drivers/clk/sunxi/Kconfig
@@ -143,4 +143,18 @@ config CLK_SUN55I_A523_R
 	  This enables common clock driver support for the PRCM
 	  in Allwinner A523/T527 SoCs.
 
+config CLK_SUN60I_A733
+	bool "Clock driver for Allwinner A733/T736"
+	default MACH_SUN60I_A733
+	help
+	  This enables common clock driver support for platforms based
+	  on Allwinner A733/T736 SoC.
+
+config CLK_SUN60I_A733_R
+	bool "Clock driver for Allwinner A733 generation PRCM"
+	default MACH_SUN60I_A733
+	help
+	  This enables common clock driver support for the PRCM
+	  in Allwinner A733/T736 SoCs.
+
 endif # CLK_SUNXI
diff --git a/drivers/clk/sunxi/Makefile b/drivers/clk/sunxi/Makefile
index 93b542cebcd..943e47612ce 100644
--- a/drivers/clk/sunxi/Makefile
+++ b/drivers/clk/sunxi/Makefile
@@ -27,3 +27,5 @@ obj-$(CONFIG_CLK_SUN50I_A64) += clk_a64.o
 obj-$(CONFIG_CLK_SUN50I_A100) += clk_a100.o
 obj-$(CONFIG_CLK_SUN55I_A523) += clk_a523.o
 obj-$(CONFIG_CLK_SUN55I_A523_R) += clk_a523_r.o
+obj-$(CONFIG_CLK_SUN60I_A733) += clk_a733.o
+obj-$(CONFIG_CLK_SUN60I_A733_R) += clk_a733_r.o
diff --git a/drivers/clk/sunxi/clk_a733.c b/drivers/clk/sunxi/clk_a733.c
new file mode 100644
index 00000000000..6ecc4bee989
--- /dev/null
+++ b/drivers/clk/sunxi/clk_a733.c
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (C) 2026 Alastair D'Silva <alastair@d-silva.org>
+ */
+
+#include <clk-uclass.h>
+#include <dm.h>
+#include <errno.h>
+#include <clk/sunxi.h>
+#include <linux/bitops.h>
+
+#include <dt-bindings/clock/sun60i-a733-ccu.h>
+#include <dt-bindings/reset/sun60i-a733-ccu.h>
+
+static struct ccu_clk_gate a733_gates[] = {
+	[CLK_PLL_PERI0_200M]	= GATE_DUMMY,
+	[CLK_APB1]		= GATE_DUMMY,
+	[CLK_APB_UART]		= GATE_DUMMY,
+
+	[CLK_SMHC0]		= GATE(0x0d00, BIT(31)),
+	[CLK_BUS_SMHC0]		= GATE(0x0d0c, BIT(0)),
+	[CLK_SMHC1]		= GATE(0x0d10, BIT(31)),
+	[CLK_BUS_SMHC1]		= GATE(0x0d1c, BIT(0)),
+	[CLK_SMHC2]		= GATE(0x0d20, BIT(31)),
+	[CLK_BUS_SMHC2]		= GATE(0x0d2c, BIT(0)),
+
+	[CLK_UART0]		= GATE(0x0e00, BIT(0)),
+	[CLK_UART1]		= GATE(0x0e04, BIT(0)),
+	[CLK_UART2]		= GATE(0x0e08, BIT(0)),
+	[CLK_UART3]		= GATE(0x0e0c, BIT(0)),
+	[CLK_UART4]		= GATE(0x0e10, BIT(0)),
+	[CLK_UART5]		= GATE(0x0e14, BIT(0)),
+	[CLK_UART6]		= GATE(0x0e18, BIT(0)),
+
+	[CLK_TWI0]		= GATE(0x0e80, BIT(0)),
+	[CLK_TWI1]		= GATE(0x0e84, BIT(0)),
+	[CLK_TWI2]		= GATE(0x0e88, BIT(0)),
+	[CLK_TWI3]		= GATE(0x0e8c, BIT(0)),
+
+	[CLK_SPI0]		= GATE(0x0f00, BIT(31)),
+	[CLK_BUS_SPI0]		= GATE(0x0f04, BIT(0)),
+	[CLK_SPI1]		= GATE(0x0f08, BIT(31)),
+	[CLK_BUS_SPI1]		= GATE(0x0f0c, BIT(0)),
+
+	[CLK_GMAC_PTP]		= GATE(0x1400, BIT(31)),
+	[CLK_GMAC0_PHY]		= GATE(0x1410, BIT(31)),
+	[CLK_GMAC0]		= GATE(0x141c, BIT(0) | BIT(1)),
+	[CLK_GMAC1_PHY]		= GATE(0x1420, BIT(31)),
+	[CLK_GMAC1]		= GATE(0x142c, BIT(0) | BIT(1)),
+
+	[CLK_USB0_OHCI]		= GATE(0x1370, BIT(31)),
+	[CLK_USB1_OHCI]		= GATE(0x1374, BIT(31)),
+	[CLK_USB0_EHCI]		= GATE(0x138c, BIT(4)),
+	[CLK_USB1_EHCI]		= GATE(0x138c, BIT(5)),
+	[CLK_USB]		= GATE(0x138c, BIT(8)),
+};
+
+static struct ccu_reset a733_resets[] = {
+	[RST_BUS_SMHC0]		= RESET(0x0d0c, BIT(16)),
+	[RST_BUS_SMHC1]		= RESET(0x0d1c, BIT(16)),
+	[RST_BUS_SMHC2]		= RESET(0x0d2c, BIT(16)),
+
+	[RST_BUS_UART0]		= RESET(0x0e00, BIT(16)),
+	[RST_BUS_UART1]		= RESET(0x0e04, BIT(16)),
+	[RST_BUS_UART2]		= RESET(0x0e08, BIT(16)),
+	[RST_BUS_UART3]		= RESET(0x0e0c, BIT(16)),
+	[RST_BUS_UART4]		= RESET(0x0e10, BIT(16)),
+	[RST_BUS_UART5]		= RESET(0x0e14, BIT(16)),
+	[RST_BUS_UART6]		= RESET(0x0e18, BIT(16)),
+
+	[RST_BUS_TWI0]		= RESET(0x0e80, BIT(16)),
+	[RST_BUS_TWI1]		= RESET(0x0e84, BIT(16)),
+	[RST_BUS_TWI2]		= RESET(0x0e88, BIT(16)),
+	[RST_BUS_TWI3]		= RESET(0x0e8c, BIT(16)),
+
+	[RST_BUS_SPI0]		= RESET(0x0f04, BIT(16)),
+	[RST_BUS_SPI1]		= RESET(0x0f0c, BIT(16)),
+
+	[RST_BUS_GMAC0]		= RESET(0x141c, BIT(16)),
+	[RST_BUS_GMAC1]		= RESET(0x142c, BIT(16)),
+
+	[RST_USB_0_OHCI]	= RESET(0x138c, BIT(16)),
+	[RST_USB_1_OHCI]	= RESET(0x138c, BIT(17)),
+	[RST_USB_0_EHCI]	= RESET(0x138c, BIT(20)),
+	[RST_USB_1_EHCI]	= RESET(0x138c, BIT(21)),
+	[RST_USB_0_DEVICE]	= RESET(0x138c, BIT(24)),
+};
+
+const struct ccu_desc a733_ccu_desc = {
+	.gates	= a733_gates,
+	.resets	= a733_resets,
+	.num_gates = ARRAY_SIZE(a733_gates),
+	.num_resets = ARRAY_SIZE(a733_resets),
+};
diff --git a/drivers/clk/sunxi/clk_a733_r.c b/drivers/clk/sunxi/clk_a733_r.c
new file mode 100644
index 00000000000..3e396e4edd1
--- /dev/null
+++ b/drivers/clk/sunxi/clk_a733_r.c
@@ -0,0 +1,38 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (C) 2026 Alastair D'Silva <alastair@d-silva.org>
+ */
+
+#include <clk-uclass.h>
+#include <dm.h>
+#include <errno.h>
+#include <clk/sunxi.h>
+#include <linux/bitops.h>
+
+#include <dt-bindings/clock/sun60i-a733-r-ccu.h>
+#include <dt-bindings/reset/sun60i-a733-r-ccu.h>
+
+static struct ccu_clk_gate a733_r_gates[] = {
+	[CLK_R_AHB]		= GATE_DUMMY,
+	[CLK_R_APBS0]		= GATE_DUMMY,
+	[CLK_R_APBS1]		= GATE_DUMMY,
+
+	[CLK_R_TWI0]		= GATE(0x19c, BIT(0)),
+	[CLK_R_TWI1]		= GATE(0x19c, BIT(1)),
+	[CLK_R_UART0]		= GATE(0x18c, BIT(0)),
+	[CLK_R_SPI]		= GATE(0x1ac, BIT(0)),
+};
+
+static struct ccu_reset a733_r_resets[] = {
+	[RST_BUS_R_TWI0]	= RESET(0x19c, BIT(16)),
+	[RST_BUS_R_TWI1]	= RESET(0x19c, BIT(17)),
+	[RST_BUS_R_UART0]	= RESET(0x18c, BIT(16)),
+	[RST_BUS_R_SPI]		= RESET(0x1ac, BIT(16)),
+};
+
+const struct ccu_desc a733_r_ccu_desc = {
+	.gates	= a733_r_gates,
+	.resets	= a733_r_resets,
+	.num_gates = ARRAY_SIZE(a733_r_gates),
+	.num_resets = ARRAY_SIZE(a733_r_resets),
+};
diff --git a/drivers/clk/sunxi/clk_sunxi.c b/drivers/clk/sunxi/clk_sunxi.c
index 842a0541bd6..ad90a3744a2 100644
--- a/drivers/clk/sunxi/clk_sunxi.c
+++ b/drivers/clk/sunxi/clk_sunxi.c
@@ -128,6 +128,8 @@ extern const struct ccu_desc r40_ccu_desc;
 extern const struct ccu_desc v3s_ccu_desc;
 extern const struct ccu_desc a523_ccu_desc;
 extern const struct ccu_desc a523_r_ccu_desc;
+extern const struct ccu_desc a733_ccu_desc;
+extern const struct ccu_desc a733_r_ccu_desc;
 
 static const struct udevice_id sunxi_clk_ids[] = {
 #ifdef CONFIG_CLK_SUN4I_A10
@@ -233,6 +235,14 @@ static const struct udevice_id sunxi_clk_ids[] = {
 #ifdef CONFIG_CLK_SUN55I_A523_R
 	{ .compatible = "allwinner,sun55i-a523-r-ccu",
 	  .data = (ulong)&a523_r_ccu_desc },
+#endif
+#ifdef CONFIG_CLK_SUN60I_A733
+	{ .compatible = "allwinner,sun60i-a733-ccu",
+	  .data = (ulong)&a733_ccu_desc },
+#endif
+#ifdef CONFIG_CLK_SUN60I_A733_R
+	{ .compatible = "allwinner,sun60i-a733-r-ccu",
+	  .data = (ulong)&a733_r_ccu_desc },
 #endif
 	{ }
 };
-- 
2.53.0

